[torch.compile] Dynamic fp8 + rms_norm fusion

ProExpertProg opened the pull request:
This PR adds support for RMSNorm + (fp8) quant fusion. It also refactors the fusion pass to make it easier to add patterns. That includes support for multiple values of epsilon.

Commit:
add fused-rms-quant-dyn-per-token branch

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Refactor fusion patterns into class

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Improved comments, move utils to MultiOutputMatch

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Allow multiple epsilons by clearing pattern matcher cache

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Add graph as property of match, add comments, add utilities, extract ops to constants

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
dynamic quant (fused ops in python)

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Upgrade reduction

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Use new dynamic ops for fusion, tolerance has to be higher.

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
In progress dynamic fusion debugging

ProExpertProg committed:
Nit comment

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Fix fusion and tests to use dynamic per-token

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Remove debug graph output

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
PR comments

Signed-off-by: luka <luka@neuralmagic.com>

mgoin commented:
@ProExpertProg It looks like there is a real failure on TPU test https://buildkite.com/vllm/fastcheck/builds/9325#01939413-9e42-4fb8-898e-dfd2a3ec7828/6-306

ProExpertProg committed:
Fix TPU test

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Abstract out quantization type to allow more quant types more easily

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
Extract getitem replacement and meta value fixing into insert_fused_node

Signed-off-by: luka <luka@neuralmagic.com>

SageMoore commented on vllm/compilation/fusion.py:
```diff
@@ -94,38 +31,593 @@ def empty_fp32(*args, **kwargs):
+        fused_node.meta["val"] = tuple(meta_val)
+
+
+class RMSNormQuantPattern:
```
How would you feel about a `QuantPattern` ABC? Looking at this PR it seems like you are extremely close to having a general purpose pattern/replacement mechanism for fusing quantize ops to any auto_functionalized op.

ProExpertProg replied:
I was going to say yes but after refactoring tuples into `QuantKey`, this class does not contain enough to be worth extracting

SageMoore commented on csrc/quantization/fused_kernels/fused_layernorm_dynamic_per_token_quant.cu:
```diff
@@ -0,0 +1,159 @@
+    std::optional<at::Tensor> scale_ub, std::optional<at::Tensor> residual) {
+  TORCH_CHECK(out.dtype() == torch::kFloat8_e4m3fn ||
+              out.dtype() == torch::kInt8);
+
```
`TORCH_CHECK(scales.dtype() == torch::kFloat32);`

SageMoore commented on csrc/quantization/fused_kernels/fused_layernorm_dynamic_per_token_quant.cu:
```diff
@@ -0,0 +1,159 @@
+          ? std::numeric_limits<float>::epsilon()
+          : 1.0f / (std::numeric_limits<c10::Float8_e4m3fn>::max() * 512.f);
+
+  if (residual.has_value()) {
```
Nit: Could you merge these and just use `residual.has_value() ? residual->data_ptr<scalar_in_t>() : nullptr`

ProExpertProg replied:
Sadly no because of the template parameter

SageMoore reviewed (commented):
Nice work. I mostly just have nits.

bnellnm commented on vllm/_custom_ops.py:
```diff
@@ -249,6 +249,40 @@ def advance_step_flashinfer(num_seqs: int, num_queries: int, block_size: int,
+    return output, scales
+
+
+if hasattr(torch.ops._C, "rms_norm_dynamic_per_token_quant"):
+
+    @register_fake("_C::rms_norm_dynamic_per_token_quant")
+    def _rms_norm_dynamic_per_token_quant_fake(
+            output: torch.Tensor,
+            input: torch.Tensor,
+            weight: torch.Tensor,
+            scales: torch.Tensor,
+            epsilon: float,
+            scale_ub: Optional[torch.Tensor] = None,
+            residual: Optional[torch.Tensor] = None) -> None:
+        return None
```
nullary ops shouldn't require fake functions.

bnellnm commented on tests/kernels/test_fused_quant_layernorm.py:
Can you add some `opcheck` test for the new op(s)?

bnellnm commented on csrc/quantization/fused_kernels/layernorm_utils.cuh:
```diff
@@ -0,0 +1,321 @@
+}
+
+// Vectorized version of vllm::compute_dynamic_per_token_scales
+// hidden_size must be a multiple of 4
```
Is there an assert for this somewhere?

ProExpertProg replied:
Yes, there's an if statement

bnellnm commented on csrc/quantization/fused_kernels/quant_conversions.cuh:
```diff
@@ -0,0 +1,79 @@
+#endif
+}
+
+#define FP8_E4M3_MAX std::numeric_limits<c10::Float8_e4m3fn>::max()
```
Can this be a constexpr?

ProExpertProg replied:
Removed, already defined

bnellnm commented on vllm/compilation/fusion.py:
```diff
@@ -94,38 +31,593 @@ def empty_fp32(*args, **kwargs):
     return ret
 
 
+class MultiOutputMatch(abc.ABC):
```
+1

ProExpertProg committed:
- OpOverload typing
- extracted MultiOutputMatch to own file
- extracted utils to fx_utils
- added named tuples for op keys

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
PR comments: kernels

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg committed:
PR comments: opcheck

Signed-off-by: luka <luka@neuralmagic.com>

tlrmchlsmth reviewed (commented):
I found some potential integer overflows. I don't think those will automatically promote to 64-bit even though the output is `int64_t`

ProExpertProg committed:
Fix dispatch utils for AMD

Signed-off-by: luka <luka@neuralmagic.com>

ProExpertProg commented on CMakeLists.txt:
```diff
@@ -196,6 +196,7 @@ set(VLLM_EXT_SRC
   "csrc/quantization/gptq/q_gemm.cu"
   "csrc/quantization/compressed_tensors/int8_quant_kernels.cu"
   "csrc/quantization/fp8/common.cu"
+  "csrc/quantization/fused_kernels/fused_layernorm_dynamic_per_token_quant.cu"
```
Clarify layernorm

ProExpertProg commented on tests/compile/test_fusion.py:
```diff
@@ -69,16 +82,28 @@ def test_fusion_rmsnorm_quant(dtype, hidden_size, num_tokens, eps):
+    fp8_quant = QUANT_OPS[key]
 
     # In pre-nodes, fp8 quant should be present and fused kernels should not
     assert find_auto_fn_maybe(pre_nodes, rms_quant) is None
```
Future PR: use compilation counters for patterns replaced

ProExpertProg replied:
(use context manager to check the counter is increased by certain number)

youkaichao reviewed (approved):
LGTM, excited to see the perf numbers!

youkaichao commented:
please also work with @tlrmchlsmth to address his comments

ProExpertProg commented on csrc/quantization/fused_kernels/fused_layernorm_dynamic_per_token_quant.cu:
```diff
@@ -0,0 +1,160 @@
+    float const* scale_ub, float const var_epsilon,
+    float const min_scaling_factor, int32_t const hidden_size,
+    scalar_t* __restrict__ residual = nullptr) {
+  // For vectorization, token_input and token_output pointers need to be
+  // aligned at 8-byte and 4-byte addresses respectively.
+  bool const can_vectorize = hidden_size % 4 == 0;
```
The code assumes contiguous so I added that

ProExpertProg committed:
PR comments:
- add kFp8Type constant for cuda/hip agnostic torch type checking
- check contiguous
- overflow
- reduce number of tests

Signed-off-by: luka <luka@neuralmagic.com>

tlrmchlsmth reviewed (approved):
Looks good to me, thank you for the great work @ProExpertProg!

ProExpertProg commented:
Seeing a 1-2% improvement in TPOT and 2-5% in TTFT.

Fused:

```

===== RUNNING neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 FOR 120 PROMPTS WITH 1 QPS =====

Namespace(backend='vllm', base_url=None, host='localhost', port=8000, endpoint='/v1/completions', dataset=None, dataset_name='sonnet', dataset_path='sonnet.txt', max_concurrency=None, model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8', tokenizer=None, best_of=1, use_beam_search=False, num_prompts=120, logprobs=None, request_rate=1.0, burstiness=1.0, seed=0, trust_remote_code=False, disable_tqdm=False, profile=False, save_result=False, metadata=None, result_dir=None, result_filename=None, ignore_eos=False, percentile_metrics='ttft,tpot,itl', metric_percentiles='99', goodput=None, sonnet_input_len=2048, sonnet_output_len=128, sonnet_prefix_len=200, sharegpt_output_len=None, random_input_len=1024, random_output_len=128, random_range_ratio=1.0, random_prefix_len=0, hf_subset=None, hf_split=None, hf_output_len=None)
Starting initial single prompt test run...
Initial test run completed. Starting main benchmark run...
Traffic request rate: 1.0
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
============ Serving Benchmark Result ============
Successful requests:                     120       
Benchmark duration (s):                  116.77    
Total input tokens:                      224758    
Total generated tokens:                  15360     
Request throughput (req/s):              1.03      
Output token throughput (tok/s):         131.55    
Total Token throughput (tok/s):          2056.42   
---------------Time to First Token----------------
Mean TTFT (ms):                          49.27     
Median TTFT (ms):                        47.41     
P99 TTFT (ms):                           76.35     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          5.38      
Median TPOT (ms):                        5.27      
P99 TPOT (ms):                           6.20      
---------------Inter-token Latency----------------
Mean ITL (ms):                           5.38      
Median ITL (ms):                         5.18      
P99 ITL (ms):                            12.97     
==================================================

===== RUNNING neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 FOR 1200 PROMPTS WITH 10 QPS =====

Namespace(backend='vllm', base_url=None, host='localhost', port=8000, endpoint='/v1/completions', dataset=None, dataset_name='sonnet', dataset_path='sonnet.txt', max_concurrency=None, model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8', tokenizer=None, best_of=1, use_beam_search=False, num_prompts=1200, logprobs=None, request_rate=10.0, burstiness=1.0, seed=0, trust_remote_code=False, disable_tqdm=False, profile=False, save_result=False, metadata=None, result_dir=None, result_filename=None, ignore_eos=False, percentile_metrics='ttft,tpot,itl', metric_percentiles='99', goodput=None, sonnet_input_len=2048, sonnet_output_len=128, sonnet_prefix_len=200, sharegpt_output_len=None, random_input_len=1024, random_output_len=128, random_range_ratio=1.0, random_prefix_len=0, hf_subset=None, hf_split=None, hf_output_len=None)
Starting initial single prompt test run...
Initial test run completed. Starting main benchmark run...
Traffic request rate: 10.0
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
============ Serving Benchmark Result ============
Successful requests:                     1200      
Benchmark duration (s):                  122.87    
Total input tokens:                      2249049   
Total generated tokens:                  153592    
Request throughput (req/s):              9.77      
Output token throughput (tok/s):         1250.00   
Total Token throughput (tok/s):          19553.81  
---------------Time to First Token----------------
Mean TTFT (ms):                          73.50     
Median TTFT (ms):                        61.19     
P99 TTFT (ms):                           218.09    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          10.78     
Median TPOT (ms):                        10.55     
P99 TPOT (ms):                           17.48     
---------------Inter-token Latency----------------
Mean ITL (ms):                           10.78     
Median ITL (ms):                         7.20      
P99 ITL (ms):                            61.43     
==================================================

===== RUNNING neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 FOR 2400 PROMPTS WITH 20 QPS =====

Namespace(backend='vllm', base_url=None, host='localhost', port=8000, endpoint='/v1/completions', dataset=None, dataset_name='sonnet', dataset_path='sonnet.txt', max_concurrency=None, model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8', tokenizer=None, best_of=1, use_beam_search=False, num_prompts=2400, logprobs=None, request_rate=20.0, burstiness=1.0, seed=0, trust_remote_code=False, disable_tqdm=False, profile=False, save_result=False, metadata=None, result_dir=None, result_filename=None, ignore_eos=False, percentile_metrics='ttft,tpot,itl', metric_percentiles='99', goodput=None, sonnet_input_len=2048, sonnet_output_len=128, sonnet_prefix_len=200, sharegpt_output_len=None, random_input_len=1024, random_output_len=128, random_range_ratio=1.0, random_prefix_len=0, hf_subset=None, hf_split=None, hf_output_len=None)
Starting initial single prompt test run...
Initial test run completed. Starting main benchmark run...
Traffic request rate: 20.0
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
============ Serving Benchmark Result ============
Successful requests:                     2400      
Benchmark duration (s):                  131.29    
Total input tokens:                      4498061   
Total generated tokens:                  307200    
Request throughput (req/s):              18.28     
Output token throughput (tok/s):         2339.90   
Total Token throughput (tok/s):          36601.09  
---------------Time to First Token----------------
Mean TTFT (ms):                          1462.86   
Median TTFT (ms):                        1547.10   
P99 TTFT (ms):                           3293.15   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          92.29     
Median TPOT (ms):                        96.82     
P99 TPOT (ms):                           104.09    
---------------Inter-token Latency----------------
Mean ITL (ms):                           92.23     
Median ITL (ms):                         38.79     
P99 ITL (ms):                            444.05    
==================================================
```

Unfused:
```

===== RUNNING neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 FOR 120 PROMPTS WITH 1 QPS =====

Namespace(backend='vllm', base_url=None, host='localhost', port=8000, endpoint='/v1/completions', dataset=None, dataset_name='sonnet', dataset_path='sonnet.txt', max_concurrency=None, model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8', tokenizer=None, best_of=1, use_beam_search=False, num_prompts=120, logprobs=None, request_rate=1.0, burstiness=1.0, seed=0, trust_remote_code=False, disable_tqdm=False, profile=False, save_result=False, metadata=None, result_dir=None, result_filename=None, ignore_eos=False, percentile_metrics='ttft,tpot,itl', metric_percentiles='99', goodput=None, sonnet_input_len=2048, sonnet_output_len=128, sonnet_prefix_len=200, sharegpt_output_len=None, random_input_len=1024, random_output_len=128, random_range_ratio=1.0, random_prefix_len=0, hf_subset=None, hf_split=None, hf_output_len=None)
Starting initial single prompt test run...
Initial test run completed. Starting main benchmark run...
Traffic request rate: 1.0
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
============ Serving Benchmark Result ============
Successful requests:                     120       
Benchmark duration (s):                  116.77    
Total input tokens:                      224758    
Total generated tokens:                  15360     
Request throughput (req/s):              1.03      
Output token throughput (tok/s):         131.54    
Total Token throughput (tok/s):          2056.34   
---------------Time to First Token----------------
Mean TTFT (ms):                          49.56     
Median TTFT (ms):                        47.19     
P99 TTFT (ms):                           81.86     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          5.47      
Median TPOT (ms):                        5.35      
P99 TPOT (ms):                           6.31      
---------------Inter-token Latency----------------
Mean ITL (ms):                           5.47      
Median ITL (ms):                         5.28      
P99 ITL (ms):                            13.13     
==================================================

===== RUNNING neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 FOR 1200 PROMPTS WITH 10 QPS =====

Namespace(backend='vllm', base_url=None, host='localhost', port=8000, endpoint='/v1/completions', dataset=None, dataset_name='sonnet', dataset_path='sonnet.txt', max_concurrency=None, model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8', tokenizer=None, best_of=1, use_beam_search=False, num_prompts=1200, logprobs=None, request_rate=10.0, burstiness=1.0, seed=0, trust_remote_code=False, disable_tqdm=False, profile=False, save_result=False, metadata=None, result_dir=None, result_filename=None, ignore_eos=False, percentile_metrics='ttft,tpot,itl', metric_percentiles='99', goodput=None, sonnet_input_len=2048, sonnet_output_len=128, sonnet_prefix_len=200, sharegpt_output_len=None, random_input_len=1024, random_output_len=128, random_range_ratio=1.0, random_prefix_len=0, hf_subset=None, hf_split=None, hf_output_len=None)
Starting initial single prompt test run...
Initial test run completed. Starting main benchmark run...
Traffic request rate: 10.0
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
============ Serving Benchmark Result ============
Successful requests:                     1200      
Benchmark duration (s):                  122.93    
Total input tokens:                      2249049   
Total generated tokens:                  153595    
Request throughput (req/s):              9.76      
Output token throughput (tok/s):         1249.48   
Total Token throughput (tok/s):          19545.23  
---------------Time to First Token----------------
Mean TTFT (ms):                          77.81     
Median TTFT (ms):                        61.64     
P99 TTFT (ms):                           273.52    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          11.30     
Median TPOT (ms):                        10.79     
P99 TPOT (ms):                           23.10     
---------------Inter-token Latency----------------
Mean ITL (ms):                           11.30     
Median ITL (ms):                         7.40      
P99 ITL (ms):                            64.54     
==================================================

===== RUNNING neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 FOR 2400 PROMPTS WITH 20 QPS =====

Namespace(backend='vllm', base_url=None, host='localhost', port=8000, endpoint='/v1/completions', dataset=None, dataset_name='sonnet', dataset_path='sonnet.txt', max_concurrency=None, model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8', tokenizer=None, best_of=1, use_beam_search=False, num_prompts=2400, logprobs=None, request_rate=20.0, burstiness=1.0, seed=0, trust_remote_code=False, disable_tqdm=False, profile=False, save_result=False, metadata=None, result_dir=None, result_filename=None, ignore_eos=False, percentile_metrics='ttft,tpot,itl', metric_percentiles='99', goodput=None, sonnet_input_len=2048, sonnet_output_len=128, sonnet_prefix_len=200, sharegpt_output_len=None, random_input_len=1024, random_output_len=128, random_range_ratio=1.0, random_prefix_len=0, hf_subset=None, hf_split=None, hf_output_len=None)
Starting initial single prompt test run...
Initial test run completed. Starting main benchmark run...
Traffic request rate: 20.0
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
============ Serving Benchmark Result ============
Successful requests:                     2400      
Benchmark duration (s):                  132.63    
Total input tokens:                      4498061   
Total generated tokens:                  307200    
Request throughput (req/s):              18.10     
Output token throughput (tok/s):         2316.27   
Total Token throughput (tok/s):          36231.34  
---------------Time to First Token----------------
Mean TTFT (ms):                          1921.30   
Median TTFT (ms):                        2041.44   
P99 TTFT (ms):                           4459.93   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          93.22     
Median TPOT (ms):                        98.07     
P99 TPOT (ms):                           104.94    
---------------Inter-token Latency----------------
Mean ITL (ms):                           93.13     
Median ITL (ms):                         39.18     
P99 ITL (ms):                            431.80    
==================================================
```
